app/flatpak-builtins-build-init.c

```diff
@@ -60,6 +60,40 @@ static GOptionEntry options[] = {
   { NULL }
 };
 
+static gboolean
+extension_matches_partial_ref (FlatpakExtension *ext, const char *match_partial_ref)
+{
+  g_autofree char *ext_arch = NULL;
+  g_autofree char *ext_branch = NULL;
+  g_autofree char *ext_base_ref = NULL;
+  g_autofree char *match_id = NULL;
+  g_autofree char *match_arch = NULL;
+  g_autofree char *match_branch = NULL;
+  g_autofree char *match_full_ref = NULL;
+
+  if (!flatpak_split_partial_ref_arg (ext->ref, FLATPAK_KINDS_RUNTIME, NULL, NULL, NULL, NULL, &ext_arch, &ext_branch, NULL))
+    return FALSE;
+
+  /* complete provided partial ref with arch and branch from the ext */
+  if (!flatpak_split_partial_ref_arg (match_partial_ref, FLATPAK_KINDS_RUNTIME, ext_arch, ext_branch, NULL, &match_id, &match_arch, &match_branch, NULL))
+    return FALSE;
+  match_full_ref = g_build_filename ("runtime", match_id, match_arch, match_branch, NULL);
+
+  if (strcmp (ext->ref, match_full_ref) == 0)
+    return TRUE;
+
+  if (strcmp (ext->id, ext->installed_id) == 0)
+    return FALSE;
```

Comment:
Could you not do this at the very top of the function?

Reply:
The first half of the function parses an extension name//maybe_ref and prints a full ref name, to compare with the full ref of the extension, returning TRUE if they match. This check is to optimise the 2nd half of the function in the case that ext->id and ext->installed_id are the same before we do more parsing/canonicalisation work to check the other one - they differ on a subdirectory extension but the first half of the function is still needed in the normal case.

Reply:
:+1: makes sense